sdk/Pulumi/Provider/ComponentAnalyzer.cs

```diff
@@ -0,0 +1,393 @@
+// Copyright 2025, Pulumi Corporation
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Reflection;
+namespace Pulumi.Experimental.Provider
+{
+    /// <summary>
+    /// Analyzes component resource types and generates a package schema.
+    /// </summary>
+    public sealed class ComponentAnalyzer
+    {
+        private readonly Metadata metadata;
+        private readonly Dictionary<string, ComplexTypeSpec> typeDefinitions = new();
+
+        private ComponentAnalyzer(Metadata metadata)
+        {
+            this.metadata = metadata;
+        }
+
+        /// <summary>
+        /// Analyzes the components in the given assembly and generates a package schema.
+        /// </summary>
+        /// <param name="metadata">The package metadata including name (required), version and display name (optional)</param>
+        /// <param name="assembly">The assembly containing component resource types to analyze</param>
+        /// <returns>A PackageSpec containing the complete schema for all components and their types</returns>
+        public static PackageSpec GenerateSchema(Metadata metadata, Assembly assembly)
+        {
+            var types = assembly.GetTypes()
+                .Where(t => typeof(ComponentResource).IsAssignableFrom(t) && !t.IsAbstract);
+            return GenerateSchema(metadata, types.ToArray());
+        }
+
+        /// <summary>
+        /// Analyzes the specified component types and generates a package schema.
+        /// </summary>
+        /// <param name="metadata">The package metadata including name (required), version and display name (optional)</param>
+        /// <param name="componentTypes">The component resource types to analyze</param>
+        /// <returns>A PackageSpec containing the complete schema for all components and their types</returns>
+        public static PackageSpec GenerateSchema(Metadata metadata, params Type[] componentTypes)
+        {
+            if (metadata?.Name == null || string.IsNullOrWhiteSpace(metadata.Name))
```

Comment:
Should we evaulate this against the same regex we have in [`pulumi.json`](https://github.com/pulumi/pulumi/blob/2348c62e6dcfbbaf737b67963546aa5110057d67/pkg/codegen/schema/pulumi.json#L11)?

Also micronit, I'd prefer curly braces in all if statements even if they are just one line.  Not sure we have a linter or guidelines for that though.